### PR TITLE
Reworked 'org.shipkit.gradle-plugin' so that it is ready for prime time

### PR DESCRIPTION
**Pull request by szczepiq**
'org.shipkit.gradle-plugin' plugin is now ready for prime time and supports release notes generation and all other magic.

- implemented handling of Gradle Plugin Portal authentication, we support env variables and project properties for those.
- ensured the right task order so the we build first, and do release magic later :)
- ensured that secret values (portal keys) are hidden from the log
- refactorings, tidy-ups, coverage, javadoc
- removed dead code

Tested by running locally for following scenarios with shipkit project iteslf. It should work. To find out if it works I need to do a real release :)

- publish plugins (with hacked version to avoid actual publication to portal), failure expected:

```
  :publishPlugins FAILED
  
  FAILURE: Build failed with an exception.
  
  * What went wrong:
  Execution failed for task ':publishPlugins'.
  > Failed to post to server.
    Server responded with:
    Plugin: 'org.shipkit.bintray', version: '0.8.114' exists already. It was created: 2017-06-21 20:09:19.674
  
  * Try:
  Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
  
  BUILD FAILED
  
  Total time: 2.57 secs


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

- perform release with dry run:

```
[./gradlew] 
[./gradlew] BUILD SUCCESSFUL
[./gradlew] 
[./gradlew] Total time: 6.388 secs

BUILD SUCCESSFUL

Total time: 7.747 secs
```

- release needed:

```
~/mockito/release$ ./gradlew releaseNeeded
  Building version '0.8.115' (value loaded from 'version.properties' file).
  'git push' does not use GitHub write token because it was not specified.
  Adding 17 discovered Gradle plugins to 'pluginBundle'
:identifyGitBranch
  Executing:
    git rev-parse --abbrev-ref HEAD
  Current branch: sf
:releaseNeeded
  Environment variable SKIP_RELEASE present: false
  Commit message to inspect for keyword '[ci skip-release]': <unknown commit message>
  Current branch 'sf' matches 'master|release/.+': false
  Release is needed: false
    - skip by env variable: false
    - skip by commit message: false
    - is pull request build:  false
    - is releasable branch:  false
    - publications changed since previous release:  true

BUILD SUCCESSFUL
```

- release notes in preview mode

```
~/mockito/release$ ./gradlew updateReleaseNotes -Ppreview
  Building version '0.8.115' (value loaded from 'version.properties' file).
  'git push' does not use GitHub write token because it was not specified.
  Adding 17 discovered Gradle plugins to 'pluginBundle'
:fetchAllContributors UP-TO-DATE
:fetchReleaseNotes UP-TO-DATE
:updateReleaseNotes
  Building new release notes based on /Users/sfaber/mockito/release/docs/release-notes.md
  Preview of release notes update:
  ----------------
**0.8.115 (2017-07-06)** - [11 commits](https://github.com/mockito/shipkit/compare/v0.8.114...v0.8.115) by [Szczepan Faber](http://github.com/szczepiq) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.8.115-green.svg)](https://plugins.gradle.org/plugin/org.shipkit.java/0.8.115)
 - No pull requests referenced in commit messages.

----------------

BUILD SUCCESSFUL
```

### CODE REVIEW

**Comment by szczepiq** [comment]
Friendly ping :) Starting from next week, my contribution rate will slow down a notch.

File: src/main/groovy/org/shipkit/internal/util/EnvVariables.java [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+
+    /**
+     * Returns env variable or null if the env variable is empty
+     */
+    public String getNonEmptyEnv(String name) {
+        String value = System.getenv(name);
+        return getNonEmpty(value);
+    }
+
+    static String getNonEmpty(String value) {
```
Comment (szpak) [inline src/main/groovy/org/shipkit/internal/util/EnvVariables.java]: Why this is `static` why the others are not?

Comment (szpak) [inline src/main/groovy/org/shipkit/internal/util/EnvVariables.java]: 2. `getNotEmptyOrNull()`?

File: src/test/groovy/org/shipkit/internal/gradle/util/StringUtilTest.groovy [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import spock.lang.Specification
+
+class StringUtilTest extends Specification {
+
+    def "knows if empty"() {
+        expect:
+        StringUtil.isEmpty(null)
+        StringUtil.isEmpty("")
+        !StringUtil.isEmpty(" ")
+        !StringUtil.isEmpty("xx")
```
Comment (szpak) [inline src/test/groovy/org/shipkit/internal/gradle/util/StringUtilTest.groovy]: Not very much in the [Spock style](http://spockframework.org/spock/docs/1.1/data_driven_testing.html) (where parameterized tests are easy to write).

File: src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    def "sets secret based on env var"() {
+        env.getNonEmptyEnv(GradlePortalReleasePlugin.PUBLISH_SECRET_ENV) >> "ksh"
+        new GradlePortalReleasePlugin(env).apply(project)
+        GradlePortalPublishTask t = project.tasks[PERFORM_PUBLISH_TASK]
+
+        expect:
+        t.publishSecret == 'ksh'
+    }
+
+    def "dry run"() {
```
Comment (szpak) [inline src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy]: You meant "should not fail in dry run"?

Comment (szpak) [inline src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy]: Or there is other purpose of that test?

File: src/test/groovy/org/shipkit/internal/gradle/AutoVersioningPluginTest.groovy [hunk 4]
```diff
@@ -1,12 +0,0 @@
-package org.shipkit.internal.gradle
-
-import testutil.PluginSpecification
-
-class AutoVersioningPluginTest extends PluginSpecification {
```
Comment (szpak) [inline src/test/groovy/org/shipkit/internal/gradle/AutoVersioningPluginTest.groovy]: 👍

**Comment by szpak** [comment]
I like it. Pure and simple.

File: src/main/groovy/org/shipkit/internal/util/EnvVariables.java (hunk 1 above)
Comment (szczepiq) [inline src/main/groovy/org/shipkit/internal/util/EnvVariables.java]: I'm ok either way. This method is not public - it is intended to help testing without actually depending on env variables.

File: src/test/groovy/org/shipkit/internal/gradle/util/StringUtilTest.groovy (hunk 2 above)
Comment (szczepiq) [inline src/test/groovy/org/shipkit/internal/gradle/util/StringUtilTest.groovy]: Very true :)

It is a very simple logic, 4 lines, I'm not convinced spock tables would add value.

File: src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy (hunk 3 above)
Comment (szczepiq) [inline src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy]: Good feedback, it is not very clear. I will update the test.

**Comment by szczepiq** [comment]
@szpak, thank you very much for comprehensive code review!

Team, given great code review, I'm merging the change as soon as the build is happy with the recent tweak to the test.

File: src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy [hunk 5]
```diff
@@ -83,11 +83,14 @@ class GradlePortalReleasePluginTest extends PluginSpecification {
...
+    def "dry run effectively disables the task"() {
         project.shipkit.dryRun = true
         project.plugins.apply(GradlePortalReleasePlugin.class)
 
-        expect:
+        when:
         project.tasks[PERFORM_PUBLISH_TASK].execute()
+
+        then:
+        noExceptionThrown() //normally the task would fail because inputs are not set
```
Comment (szpak) [inline src/test/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePluginTest.groovy]: 👍

**Pull request merged**